### PR TITLE
Replace D1 operator overloads in SuspendableThrottlerCount

### PR DESCRIPTION
**Pull request by joseph-wakeling-frequenz**
DMD 2.088.0 has deprecated D1 operator overloads, which are expected to be replaced with the newer D2 `op{Unary,Binary,BinaryRight,OpAssign}` templated methods: https://dlang.org/changelog/2.088.0.html#dep_d1_ops

`SuspendableThrottlerCount` has a slightly finnicky implementation of these, so it seems worth updating it in a separate patch.  Its original design defined local methods (`inc`/`dec`, `add`/`remove`) that provide the actual implementation, and used aliases to these to implement the actual operator overloads.  This is a problem because these methods may be (and in at least some test cases are) overridden by subclasses.  We therefore cannot just replace them with standard D2 operator overloads, as templated methods cannot be overridden by subclasses.

The simplest solution therefore seems to be to continue to support these custom local methods, and replace each D1 operator overload alias with a D2 operator overload implemented as a thin wrapper of the local method.

This should allow overloads of the implementations to continue to work, while having no impact on subclasses or users of the operators.

Part of https://github.com/sociomantic-tsunami/ocean/issues/747.

### CODE REVIEW

**Comment by joseph-wakeling-frequenz** [comment]
Updated with a cross-ref to the issue, which I hadn't seen before submitting.

**Comment by ben-palmer-sociomantic** [comment]
Should we take the same approach here as we did in https://github.com/sociomantic-tsunami/ocean/pull/749 ?

**Comment by joseph-wakeling-frequenz** [comment]
I don't think so, because these seem like reasonable use-cases for the operators -- we're literally incrementing or decrementing a count -- and unlike that PR, these are "in-place" updates.

Unless anyone has a different opinion?

**Comment by ben-palmer-sociomantic** [comment]
LGTM then, maybe we should have some release notes though. @don-clugston-sociomantic could you maybe have a look at this as well?

**Comment by joseph-wakeling-frequenz** [comment]
> LGTM then, maybe we should have some release notes though.

I was planning on doing one big release-notes section for replacement of operator overloads, after everything was done.  Is that OK or would you prefer me to go module by module?

**Comment by ben-palmer-sociomantic** [comment]
> Is that OK or would you prefer me to go module by module?

Hmm, up to you. Might be easier to manage having every PR independent.

**Comment by joseph-wakeling-frequenz** [comment]
> Hmm, up to you. Might be easier to manage having every PR independent.

OK.  I'll write up release notes for this patch covering the general principle of replacing D1 operator overloads, and update them module by module.

**Comment by ben-palmer-sociomantic** [comment]
Just waiting on the release notes for this before merging.

**Comment by joseph-wakeling-frequenz** [comment]
Oh, sorry!  I'll try to get that done later today.

**Comment by joseph-wakeling-frequenz** [comment]
Release notes added :-)  I will continue to extend these in future PRs.

**Comment by ben-palmer-sociomantic** [comment]
Cool, looks good. Just one thing that occurred to me. I think it would be worth adding a short unit test for these changes.

**Comment by joseph-wakeling-frequenz** [comment]
It's good that you suggested that, because it turns out there is a bug in the original `opSubAssign`, which aliases to `add` instead of the correct `remove` (!).

I guess that means that probably no one is actually using that operator :-P

**Comment by joseph-wakeling-frequenz** [comment]
I've pushed a bugfix patch with the operator overload migration rebased on top.  How would you like to handle this?  Would you prefer that I submit the bugfix patch against v5.1.x first ... ?

**Comment by joseph-wakeling-frequenz** [comment]
Updated with a rewritten commit message and release notes for the bugfix patch: I realized it's probably best to describe the fix in terms of the operator rather than `opSubAssign`. since we're getting rid of the latter.

**Comment by ben-palmer-sociomantic** [comment]
> Would you prefer that I submit the bugfix patch against v5.1.x first ... ?

Yeah I think that would be the best option. Then we can also merge the original changes in to v5.x.x

**Comment by joseph-wakeling-frequenz** [comment]
Here you go: https://github.com/sociomantic-tsunami/ocean/pull/771 :-)

**Comment by joseph-wakeling-frequenz** [comment]
Rebased on latest v5.x.x branch.  This should be good to go, I think ... ?

**Review by ben-palmer-sociomantic — APPROVED** [review]
LGTM

**Pull request merged**
